actionpack/lib/action_controller/metal/renderers.rb

```diff
@@ -18,16 +18,33 @@ def initialize(format)
     end
   end
 
+  class MissingSerializingFormat < KeyError
+    def initialize(format)
+      super "No serializing format defined for: '#{format}'."
+    end
+  end
+
   module Renderers
     extend ActiveSupport::Concern
 
     # A Set containing renderer names that correspond to available renderer procs.
     # Default values are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
     RENDERERS = Set.new
+    FORMATTERS = Hash.new do |h, format|
```

Comment:
In any given controller, there can be one and only one formatter for the `:json` format.

In general, the keys of the FORMATTERS Hash should always equal the elements in the RENDERERS Set (which is a Hash under the covers, anyhow).  

Or do you see something here I don't?

I do think there should be some refactoring to follow this up, but given all the back and forth discussion, chose the least-invasive implementation.


tl;dr summary of changes

```diff
# module Renderers

   included do
       class_attribute :_renderers
       self._renderers = Set.new.freeze
+      class_attribute :_formatters
+      self._formatters = FORMATTERS.dup.freeze
+    end

class_methods do
+      def serializing(**formatters)
+        self._formatters = _formatters.merge(formatters)
+      end
```

```diff
# module Renderers::All

      included do
         self._renderers = RENDERERS
+        self._formatters = FORMATTERS
       end
```

```diff
# module ActionController
+    def self.add_formatter(format, &block)
+      FORMATTERS[format.to_sym] = block
+    end
+

+    add_formatter :json do |json, options|
+      json.kind_of?(String) ? json : json.to_json(options)
+    end
     add :json do |json, options|
-      json = json.to_json(options) unless json.kind_of?(String)
+      json = _formatters[:json].call(json, options)
```